File: web/processor-app/src/types/trade.ts
import { z } from "zod";
import { Datetime } from "./primitives/datetime";
import { Fraction } from "./primitives/fraction";
import { Direction } from "./primitives/direction";
import { Uuid } from "./primitives/uuid";

export const PublicTrade = z.object({
  price: Fraction,
  volume: Fraction,
  time: Datetime,
  direction: Direction,
});
export type PublicTrade = z.infer<typeof PublicTrade>;

export const PrivateTrade = z.object({
  id: Uuid,
  created_at: Datetime,
  quote_asset_id: Uuid,
  base_asset_id: Uuid,
  price: Fraction,
  taker_quote_volume: Fraction,
  taker_base_volume: Fraction,
  maker_quote_volume: Fraction,
  maker_base_volume: Fraction,
  direction: Direction,
});
export type PrivateTrade = z.infer<typeof PrivateTrade>;

File: server/engagement/src/database/projections/trade.rs
use chrono::{DateTime, Utc};
use fraction::Fraction;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Trade {
    pub id: Uuid,
    pub created_at: DateTime<Utc>,
    pub last_modification_at: DateTime<Utc>,
    pub quote_asset_id: Uuid,
    pub base_asset_id: Uuid,
    pub taker_id: Uuid,
    pub taker_presentation: bool,
    pub maker_id: Uuid,
    pub maker_presentation: bool,
    pub price: Fraction,
    pub taker_quote_volume: Fraction,
    pub taker_base_volume: Fraction,
    pub maker_quote_volume: Fraction,
    pub maker_base_volume: Fraction,
}
